[RFC] Retry fgets on EINTR (#7632)

mattwidmann opened the pull request:
This fixes #7632 for me. It's tricky to test, since it relies on a race between a subprocess exiting and waiting for cscope's output.

- Should we instead change all of these to use `vim_fgets`? I didn't here because `vim_fgets` has some additional handling for lines that are too long, but that might be alright for all or some of these cases.
- The linter is complaining about missing braces on if statements I didn't change -- is that expected?

justinmk commented:
Thanks!

> Should we instead change all of these to use vim_fgets

Better to change one thing at a time.

> The linter is complaining about missing braces on if statements I didn't change -- is that expected?

Yes, it "cascades" to nearby lines, so legacy code gets updated incrementally.

justinmk commented on src/nvim/fileio.c:
```diff
@@ -4448,14 +4448,24 @@ bool vim_fgets(char_u *buf, int size, FILE *fp) FUNC_ATTR_NONNULL_ALL
+      errno = 0;
       ignoredp = fgets((char *)tbuf, FGETS_SIZE, fp);
+      if (ignoredp == NULL && errno == EINTR) {
+        goto retry_ignore;
```
technically we're not ignoring the retval now, so `ignoredp` should not be used.

mattwidmann replied:
Yeah, I guess we can just rely on the existing loop to retry the `fgets` here.

mattwidmann replied:
Wait, this won't properly handle an EOF, either...

justinmk replied:
just to be clear, i was only speaking about the specific `ignoredp` global which is used as a dummy assignment when Vim doesn't care about some return value.

justinmk commented:
would be good to include your explanation and debugging steps from https://github.com/neovim/neovim/issues/7632 , in the commit message.

Commit:
io: fix handling EOF in vim_fgets

If an EOF is returned from `fgets`, `vim_fgets` might spin forever, as
it tries to consume the current line.

A `NULL` return value from `fgets` should break out of the function
(unless `errno` is `EINTR`), and then `feof` should be used to check for
the EOF condition on the stream.

mattwidmann commented:
I had to rewrite the `vim_fgets` function to make it safe (could just spin forever if EOF is returned from one of the calls to `fgets`. I could put this in a separate commit when I squash, since it's a separate (though overlapping) fix.

I also pushed a linter fix and an empty commit with the proposed commit message.

justinmk commented:
LGTM, just leave a comment when ready.

Commit:
io: retry fgets on EINTR (#7632)

The calls to `fgets` in `src/nvim/if_cscope.c` (and elsewhere) can show
communication errors to the user if a signal is delivered during its
system calls. For plugins that proxy subprocess output into cscope
requests, a `SIGCHLD` might *always* interfere with calls into `fgets`.

To see this in a debugger, put a breakpoint on `cs_reading_emsg` and
watch signals come in (with lldb, using `process handle --notify true
--pass true`).  Next, run a subcommand from neovim that calls through
cscope when it returns.  A tag picker plugin, like vim-picker and fzy,
with `cscopetag` and `cscopetagorder=0` set, reproduced this reliably.
The breakpoint will hit after a `SIGCHLD` is delivered, and `errno` will
be set to 4, `EINTR`.

The caller of `fgets` should retry when `NULL` is returned with `errno`
set to `EINTR`.

mattwidmann commented:
Looking at the test failure here: <http://neovim-qb.szakmeister.net/build/15079/tap_report/by_test?reportset=functionaltest-freebsd-64-debug>

Is this test flaky? This change could have caused issues here, but the test looks somewhat racy. The failure is:

```
test/functional/eval/timer_spec.lua @ 41
Failure message: test/functional/eval/timer_spec.lua:44: Expected objects to be the same.
Passed in:
(number) 1
Expected:
(number) 0
stack traceback:
test/functional/eval/timer_spec.lua:44: in function <test/functional/eval/timer_spec.lua:41>

test/functional/eval/timer_spec.lua @ 49: timers works with zero timeout
```

And the test is:

```lua
  it('are triggered during sleep', function()
    command("call timer_start(50, 'MyHandler', {'repeat': 2})")
    nvim_async("command", "sleep 10")
    eq(0,eval("g:val"))
    run(nil, nil, nil, 300)
    eq(2,eval("g:val"))
  end)
```

There's nothing preventing the `timer_start` command and the check that `g:val` equals 0 from being separated by 50 ms or more. This didn't reproduce on the other test harnesses, either.

justinmk commented:
Yes, that test is flaky. Can ignore it here.

mattwidmann commented:
Alright -- I've pushed two commits here and tested them locally. One of the fileio tests is failing on 32-bit:

```
[32m[2m[ RUN      ][0m[0m file_read can read small chunks of input until eof: 
```

That's an alarming test to have fail against this branch. I didn't change that fileio code -- it uses its own buffering scheme instead of `FILE` streams, so it makes the read or readv syscalls itself.

Have you seen that test fail before? Is there any way I can test 32-bit on macOS? I ran that test on my machine and it passed, though it took about a second to complete. Is it timing out, maybe?

justinmk commented:
@mattwidmann That is a very weird failure, it looks like a timeout (build log stopped abruptly), but travis didn't flag it as a timeout. Have not seen that before.

I restarted the test since the logs don't show anything useful,  to see if the failure is consistent.

mattwidmann commented:
Looks like the build failed this time, trying to fetch dependencies:

```E: Failed to fetch http://ppa.launchpad.net/ubuntu-toolchain-r/test/ubuntu/pool/main/g/gcc-5/g++-5-multilib_5.4.1-2ubuntu1~14.04_amd64.deb  Unable to connect to ppa.launchpad.net:http:
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
apt-get.diagnostics
apt-get install failed
[...]
The command "sudo -E apt-get -yq --no-install-suggests --no-install-recommends --force-yes install autoconf automake apport build-essential clang-4.0 cmake cscope g++-5-multilib g++-multilib gcc-5-multilib gcc-multilib gdb language-pack-tr libc6-dev-i386 libtool llvm-4.0-dev locales pkg-config unzip valgrind xclip" failed and exited with 100 during .
```

Might need to be restarted again.

justinmk commented:
restarted again. definitely looks like travis issue:

```
oci runtime error: exec failed: container_linux.go:265: starting container process caused "could not create session key: disk quota exceeded"
```

mattwidmann commented:
Well, I'm done with this PR -- feel free to merge it when you're ready. Let me know if anything should be changed.
